File: Benchmarking/Swift/matinv.swift
import Foundation

func readMatrix(from filename: String) -> [[Double]]? {
    guard let fileHandle = FileHandle(forReadingAtPath: filename) else {
        print("Error opening file")
        return nil
    }
    
    let data = fileHandle.readDataToEndOfFile()
    guard let content = String(data: data, encoding: .utf8) else {
        print("Error reading file content")
        return nil
    }
    
    let lines = content.split(separator: "\n").map { $0.trimmingCharacters(in: .whitespacesAndNewlines) }
    guard let n = Int(lines[0]) else {
        print("Error parsing matrix size")
        return nil
    }
    
    var matrix = [[Double]](repeating: [Double](repeating: 0.0, count: n), count: n)
    
    for i in 0..<n {
        let row = lines[i + 1].split(separator: " ").compactMap { Double($0) }
        if row.count == n {
            matrix[i] = row
        } else {
            print("Error parsing matrix row")
            return nil
        }
    }
    
    return matrix
}

func writeMatrix(to filename: String, matrix: [[Double]]) {
    let n = matrix.count
    var content = ""
    
    for i in 0..<n {
        let row = matrix[i].map { String($0) }.joined(separator: " ")
        content += row + "\n"
    }
    
    do {
        try content.write(toFile: filename, atomically: true, encoding: .utf8)
    } catch {
        print("Error writing file: \(error)")
    }
}

func gaussJordanInverse(matrix: [[Double]]) -> [[Double]]? {
    let n = matrix.count
    var augmented = [[Double]](repeating: [Double](repeating: 0.0, count: 2 * n), count: n)
    
    // Create augmented matrix with the identity matrix on the right
    for i in 0..<n {
        for j in 0..<n {
            augmented[i][j] = matrix[i][j]
            augmented[i][j + n] = (i == j) ? 1.0 : 0.0
        }
    }
    
    // Perform Gauss-Jordan elimination
    for i in 0..<n {
        var pivot = augmented[i][i]
        if pivot == 0.0 {
            var swapRow = -1
            for j in (i + 1)..<n {
                if augmented[j][i] != 0.0 {
                    swapRow = j
                    break
                }
            }
            if swapRow == -1 {
                print("Matrix is singular and cannot be inverted.")
                return nil
            }
            
            // Swap rows
            augmented.swapAt(i, swapRow)
            pivot = augmented[i][i]
        }
        
        // Normalize the pivot row
        for j in 0..<2 * n {
            augmented[i][j] /= pivot
        }
        
        // Eliminate the current column in all other rows
        for j in 0..<n {
            if j != i {
                let factor = augmented[j][i]
                for k in 0..<2 * n {
                    augmented[j][k] -= factor * augmented[i][k]
                }
            }
        }
    }
    
    // Extract the inverse matrix from the augmented matrix
    var inverse = [[Double]](repeating: [Double](repeating: 0.0, count: n), count: n)
    for i in 0..<n {
        for j in 0..<n {
            inverse[i][j] = augmented[i][j + n]
        }
    }
    
    return inverse
}

// Main function
func main() {
    guard let matrix = readMatrix(from: "../random_matrix.txt") else {
        return
    }
    
    guard let inverse = gaussJordanInverse(matrix: matrix) else {
        return
    }
    
    writeMatrix(to: "./inverse_matrix.txt", matrix: inverse)
}

// Call main
main()

File: Benchmarking/Perl/matinv.pl
use strict;
use warnings;
use File::Slurp;

sub read_matrix {
    my ($filename) = @_;
    my @matrix;
    
    my $data = read_file($filename);
    my @lines = split /\n/, $data;
    my $n = shift @lines;
    
    foreach my $line (@lines) {
        my @row = split ' ', $line;
        push @matrix, \@row;
    }
    
    return ($n, \@matrix);
}

sub write_matrix_to_file {
    my ($filename, $matrix) = @_;
    my $n = scalar @$matrix;
    my @lines;
    
    foreach my $row (@$matrix) {
        push @lines, join(' ', @$row);
    }
    
    write_file($filename, join("\n", @lines) . "\n");
}

sub gauss_jordan_inverse {
    my ($matrix, $n) = @_;
    my @augmented;
    
    # Create an augmented matrix with the identity matrix on the right
    for my $i (0..$n-1) {
        push @augmented, [ @{$matrix->[$i]}, (0) x $n ];
        $augmented[$i][$i + $n] = 1;
    }
    
    # Perform Gauss-Jordan elimination
    for my $i (0..$n-1) {
        my $pivot = $augmented[$i][$i];
        if ($pivot == 0) {
            my $swap_row = -1;
            for my $j ($i + 1 .. $n - 1) {
                if ($augmented[$j][$i] != 0) {
                    $swap_row = $j;
                    last;
                }
            }
            die "Matrix is singular and cannot be inverted.\n" if $swap_row == -1;
            
            # Swap rows
            @augmented[$i, $swap_row] = @augmented[$swap_row, $i];
            $pivot = $augmented[$i][$i];
        }
        
        # Normalize the pivot row
        for my $j (0 .. 2 * $n - 1) {
            $augmented[$i][$j] /= $pivot;
        }
        
        # Eliminate the current column in all other rows
        for my $j (0 .. $n - 1) {
            if ($j != $i) {
                my $factor = $augmented[$j][$i];
                for my $k (0 .. 2 * $n - 1) {
                    $augmented[$j][$k] -= $factor * $augmented[$i][$k];
                }
            }
        }
    }
    
    # Extract the inverse matrix from the augmented matrix
    my @inverse;
    for my $i (0..$n-1) {
        push @inverse, [ @{$augmented[$i]}[$n..2*$n-1] ];
    }    
    return \@inverse;
}

sub main {
    my ($n, $matrix) = read_matrix("../random_matrix.txt");
    
    my $inverse = gauss_jordan_inverse($matrix, $n);
    
    write_matrix_to_file("./inverse_matrix.txt", $inverse);
}

main();